File: ios/auto-mobile-sdk/Sources/AutoMobileSDK/Events/DropCounter.swift
import Foundation

/// Reason an event was dropped by the SDK.
public enum DropReason: String, Codable, Sendable, CaseIterable {
    case disabled
    case shutdown
    case flushError
    case bufferOverflow
    case filtered
    case deliveryFailed
}

/// Tracks dropped event counts by reason.
protocol DropCounting: AnyObject, Sendable {
    func increment(_ reason: DropReason)
    func increment(_ reason: DropReason, count: Int)
    func snapshot() -> [DropReason: Int]
    func reset()
}

extension DropCounting {
    func increment(_ reason: DropReason, count: Int) {
        for _ in 0..<count {
            increment(reason)
        }
    }
}

/// Thread-safe default implementation of ``DropCounting``.
final class DefaultDropCounter: DropCounting, @unchecked Sendable {
    private let lock = NSLock()
    private var counts: [DropReason: Int] = [:]

    init() {}

    func increment(_ reason: DropReason) {
        lock.lock()
        counts[reason, default: 0] += 1
        lock.unlock()
    }

    func increment(_ reason: DropReason, count: Int) {
        lock.lock()
        counts[reason, default: 0] += count
        lock.unlock()
    }

    func snapshot() -> [DropReason: Int] {
        lock.lock()
        defer { lock.unlock() }
        return counts
    }

    func reset() {
        lock.lock()
        counts.removeAll()
        lock.unlock()
    }
}

File: android/auto-mobile-sdk/src/main/kotlin/dev/jasonpearson/automobile/sdk/events/DropCounter.kt
package dev.jasonpearson.automobile.sdk.events

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/** Reason an event was dropped by [SdkEventBuffer]. */
enum class DropReason {
  DISABLED,
  SHUTDOWN,
  FLUSH_ERROR,
  BUFFER_OVERFLOW,
  FILTERED,
  DELIVERY_FAILED,
  PROCESSOR_ERROR,
}

/** Tracks counts of dropped events by reason. */
internal interface DropCounter {
  fun increment(reason: DropReason, count: Int = 1)

  fun snapshot(): Map<DropReason, Long>

  fun reset()
}

/** Thread-safe [DropCounter] backed by [ConcurrentHashMap] and [AtomicLong]. */
internal class DefaultDropCounter : DropCounter {

  private val counts = ConcurrentHashMap<DropReason, AtomicLong>()

  override fun increment(reason: DropReason, count: Int) {
    counts.computeIfAbsent(reason) { AtomicLong(0) }.addAndGet(count.toLong())
  }

  override fun snapshot(): Map<DropReason, Long> =
      counts.mapValues { (_, v) -> v.get() }.filterValues { it > 0 }

  override fun reset() {
    counts.values.forEach { it.set(0) }
  }
}
